### PR TITLE
Improve throughput through sharding fingerprints

### LINKED ISSUE #3093: The throughput of ingester causes a lot of remote write latency
I used cortex in production environment, which was stable, but there was a lot of latency after adding some new remote writes from Prometheus.

I use the following expression to calculate the delay of remote write:
```
(time() - prometheus_remote_storage_queue_highest_sent_timestamp_seconds{instance_name=~"$instance_name"}) < 100000
```
It's measured in seconds, and it looks terrible:
image

Through monitoring, I have good reasons to believe that Distributor and Ingester have not encountered resource bottlenecks (CPU/Memory)
Through the distributor's log, I can see that a large number of push requests take more than 5 seconds, or even more than 40 seconds, which makes me surprised. As far as I know, when the ingester receives the data, it only writes to the memory without any IO operation. Data persistence is done periodically and asynchronously. It should be very fast。
I conducted a detailed time-consuming analysis (by recording the time-consuming of the core steps in the context and passing them layer by layer). Finally, I found that 90% of the processing time of the push request (for example, 8s out of 10s) is generated in the following code:
https://github.com/cortexproject/cortex/blob/master/pkg/ingester/index/index.go#L118-L151

The time spent in this code is mainly concentrated in two places:
1. `Lock`
https://github.com/cortexproject/cortex/blob/master/pkg/ingester/index/index.go#L119
2. `Copy`
https://github.com/cortexproject/cortex/blob/master/pkg/ingester/index/index.go#L144

According to actual measurement, a certain Push takes 30s (including 10000 series), among which `Copy` takes `14427ms` and `Lock` takes `15313ms`.

As a time series database, throughput is an extremely core indicator, this problem should be solved.

---
My remote write rate is 830000 timeseries/s, each series contains only one sample and about 10 labels in every request.

### PR DESCRIPTION
**Pull request by storyicon**
**What this PR does**:
This PR is expected to increase the throughput of ingester(chunk storage) by 10-100 times.

**Which issue(s) this PR fixes**:
Fixes #3093

According to the issue description, `Lock` and `Append` take up 95% of the entire `Push` request time.
The time consumption of the `Lock` is affected by the amount of concurrency, but in essence it is affected by the time consumption of the code area it protects.
So the key is to solve the problem that `Copy` takes too long. In fact, this is a truth that any data structure beginner understands: the cost of random insertion into the array is expensive because it requires moving a large number of elements.
To be honest, I first thought of skiplist, because the scene here is very similar to redis's sorted set. After some comparative tests, the random insertion speed of skiplist is indeed much better than that of array. But when it was actually integrated into cortex and used in my environment, it took up too much memory. So I finally went back to the array.
What I consider is to divide a big ordered slice into multiple ordered slices to reduce the cost of `Copy`. It sounds a bit simple, but is really effective and suitable

Its general idea is as follows:
It divides `math.MaxUint64` (this is the value range of fingerprint) into `N` shards:
```
shards[0]: [0, math.MaxUint64 * 1/N)
shards[1]: [math.MaxUint64 * 1/N, math.MaxUint64 * 2/N)
...
shards[n]: [math.MaxUint64 * (N-1)/N, math.MaxUint64]
```
(the actual shard is not as accurate as the above).

`Fingerprints` in each shard are kept in order. In this way, we can deal with it as we did with a large array of fingerprints. It brings a small cost (shard), but greatly improves the throughput, and it even helps to reduce memory consumption to some extent (large arrays take up more memory when growing).

**Comparison**:
image

In my test, each consumer can only consume up to  `25000 series/s` before the modification, and it can consume up to `1400000 series/s` after the modification. You may have doubts about the data, I will provide the testing process below.

### CODE REVIEW

**Comment by storyicon** [comment]
When I was testing, I created a structure `InvertedIndexV2` for the modified `InvertedIndex` and made it initialized by `NewV2`.

In order to avoid other interference, I directly modified the Push method of Distributor:
```
// deduplicator, map[model.Fingerprint]struct{}{}
var fpMap sync.Map
var invertedIndex = index.New() // When testing the effect after shard, change it to index.NewV2
func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*client.WriteResponse, error) {
	for _, ts := range req.Timeseries {
		fp := client.FastFingerprint(ts.Labels)
		_, exists := fpMap.LoadOrStore(fp, struct{}{})
		if !exists {
			invertedIndex.Add(ctx, ts.Labels, fp)
		}
	}
        return &client.WriteResponse{}, nil
}
```

Below is the magnitude of my data:
1. Normally, it's `83w` timeseries/s. 
2. Each series contains only one sample and about 10 labels.
3. The cardinality of some labels is relatively large.
4. Prometheus has a lot of accumulation(latency) when testing (To test the maximum throughput capacity).

The data is not extreme, they come from our real production environment (from a large number of node_exporters).

**Comment by storyicon** [comment]
Here's a simple test out of the box:

```
package main

import (
	"fmt"
	"math"
	"math/rand"
	"sort"
	"time"

	"github.com/prometheus/common/model"
)

func GenRandomFingerprint(size int) []model.Fingerprint {
	fps := make([]model.Fingerprint, size)
	for i := 0; i < size; i++ {
		fps[i] = model.Fingerprint(rand.Uint64())
	}
	return fps
}

type IndexValueEntry struct {
	fps []model.Fingerprint
}

func NewV1() *IndexValueEntry {
	return &IndexValueEntry{}
}

func (s *IndexValueEntry) add(fp model.Fingerprint) {
	j := sort.Search(len(s.fps), func(i int) bool {
		return s.fps[i] >= fp
	})
	s.fps = append(s.fps, 0)
	copy(s.fps[j+1:], s.fps[j:])
	s.fps[j] = fp
}

type IndexValueEntryV2 struct {
	shards [][]model.Fingerprint
}

func NewV2() *IndexValueEntryV2 {
	shards := make([][]model.Fingerprint, 200)
	return &IndexValueEntryV2{
		shards: shards,
	}
}

func (s *IndexValueEntryV2) add(fp model.Fingerprint) {
	num := int(math.Floor(float64(len(s.shards)) * float64(fp) / math.MaxUint64))
	fps := s.shards[num]
	j := sort.Search(len(fps), func(i int) bool {
		return fps[i] >= fp
	})
	fps = append(fps, 0)
	copy(fps[j+1:], fps[j:])
	fps[j] = fp
	s.shards[num] = fps
}

type Indexer interface {
	add(fp model.Fingerprint)
}

func TestIndexer(indexer Indexer, initSize int, insertSize int) time.Duration {
	fps := GenRandomFingerprint(initSize)
	for _, fp := range fps {
		indexer.add(fp)
	}

	fps = GenRandomFingerprint(insertSize)
	return Elapsed(func() {
		for _, fp := range fps {
			indexer.add(fp)
		}
	})
}

func Elapsed(fn func()) time.Duration {
	t := time.Now()
	fn()
	return time.Since(t)
}

func main() {
	var (
		initSize   = 100000
		insertSize = 10000
	)
	fmt.Println("V2:", TestIndexer(NewV2(), initSize, insertSize))
	fmt.Println("V1:", TestIndexer(NewV1(), initSize, insertSize))
}
```

It prints these as it runs on my machine:
```
V2: 3.6664ms
V1: 90.6585ms
```

When you properly increase `initSize` and `insertSize`, you will find that the difference between `V1` and `V2` will be more than `30` times. At the same time, `V2` takes up less memory.
In fact, I think it's normal to have a cardinality greater than `10w`. At least in our environment, there are many cardinality over `100w`.

---

In fact, I noticed a shard here:
```
const indexShards = 32
type InvertedIndex struct {
	shards []indexShard
}
```

It can effectively reduce lock contention (I call it lock segmentation), but when the value of  `indexShards` is large, it will actually reduce the query efficiency, so it does not conflict with this pr.

**Review by pstibrany — COMMENTED** [review]
Thank you, that looks really interesting. I have left some comments in the code.

Would it be possible to write a benchmark with adding same number of metrics and compare values before and after this change? (I see you already wrote a comparison above... can you add it as a benchmark to the PR please?)

File: pkg/ingester/index/index.go [hunk 1]
```diff
@@ -88,12 +89,67 @@ func (ii *InvertedIndex) Delete(labels labels.Labels, fp model.Fingerprint) {
...
+func (c *indexValueEntry) length() int {
+	var i int
+	for _, shard := range c.shards {
+		i += len(shard)
+	}
+	return i
+}
+
+func (c *indexValueEntry) shard(fp model.Fingerprint) int {
+	return int(math.Floor(float64(len(c.shards)) * float64(fp) / math.MaxUint64))
```
[the remark this answers is not in the document]
Comment (storyicon) [reply pkg/ingester/index/index.go:139]: Thanks for correcting, I made a low-level mistake. This fully illustrates the importance of CodeReview.
```
Perhaps simplest solution would be to use fp % len(c.shards) as a shard number. (That would require little bit of extra effort when iterating them in sorted order)
```
Breaking the order of the shard will bring additional costs. I think the simplest way is this:
```
func (c *indexValueEntry) shard(fp model.Fingerprint) int {
	n := int(math.Floor(float64(len(c.shards)) * float64(fp) / math.MaxUint64))
	if n == len(c.shards) {
		n = len(c.shards) - 1
	}
	return n
}
```
What is your opinion?

File: pkg/ingester/index/index.go [hunk 2]
```diff
@@ -88,12 +89,67 @@ func (ii *InvertedIndex) Delete(labels labels.Labels, fp model.Fingerprint) {
...
+	return fps
+}
+
+func (c *indexValueEntry) delete(fp model.Fingerprint) {
+	num := c.shard(fp)
+	fps := c.shards[num]
+	j := sort.Search(len(fps), func(i int) bool {
+		return fps[i] >= fp
+	})
+	c.shards[num] = fps[:j+copy(fps[j:], fps[j+1:])]
```
[the remark this answers is not in the document]
Comment (storyicon) [reply pkg/ingester/index/index.go:124]: Indeed, I copied the previous code directly. Now I've added some conditional statements here:
```
if len(fps) == j {
    return
}
```

File: pkg/ingester/index/index.go [hunk 3]
```diff
@@ -88,12 +89,67 @@ func (ii *InvertedIndex) Delete(labels labels.Labels, fp model.Fingerprint) {
...
+func newIndexValueEntry(value string) *indexValueEntry {
+	shards := make([][]model.Fingerprint, indexValueShards)
+	return &indexValueEntry{
+		value:  value,
+		shards: shards,
+	}
+}
+
+func (c *indexValueEntry) fps() []model.Fingerprint {
+	var fps []model.Fingerprint
```
[the remark this answers is not in the document]
Comment (storyicon) [reply pkg/ingester/index/index.go:111]: that's a good idea

File: pkg/ingester/index/index.go [hunk 4]
```diff
@@ -88,12 +89,67 @@ func (ii *InvertedIndex) Delete(labels labels.Labels, fp model.Fingerprint) {
...
+}
+
+func (c *indexValueEntry) add(fp model.Fingerprint) {
+	num := c.shard(fp)
+	fps := c.shards[num]
+	// Insert into the right position to keep fingerprints sorted
+	j := sort.Search(len(fps), func(i int) bool {
+		return fps[i] >= fp
+	})
+	fps = append(fps, 0)
```
[the remark this answers is not in the document]
Comment (storyicon) [reply pkg/ingester/index/index.go:149]: 上层逻辑确保了这里不会有重复的fingerprint：

[the remark this answers is not in the document]
Comment (storyicon) [reply pkg/ingester/index/index.go:149]: The upper-level logic ensures that there will be no duplicate fingerprints here:
https://github.com/cortexproject/cortex/blob/master/pkg/ingester/user_state.go#L178-L181

**Comment by pracucci** [comment]
> Would it be possible to write a benchmark with adding same number of metrics and compare values before and after this change?

I also would be very interested into this, benchmarking both CPU and memory. For example, the following change may have an impact:
```diff
-	fps  map[string]indexValueEntry
+	fps  map[string]*indexValueEntry
```

**Comment by storyicon** [comment]
> > Would it be possible to write a benchmark with adding same number of metrics and compare values before and after this change?
> 
> I also would be very interested into this, benchmarking both CPU and memory. For example, the following change may have an impact:
> 
> ```diff
> -	fps  map[string]indexValueEntry
> +	fps  map[string]*indexValueEntry
> ```

Hmmm, the benchmark I will commit is not affected by this

**Comment by storyicon** [comment]
> Thank you, that looks really interesting. I have left some comments in the code.
> 
> Would it be possible to write a benchmark with adding same number of metrics and compare values before and after this change? (I see you already wrote a comparison above... can you add it as a benchmark to the PR please?)

I submitted a modification, it contains a simple benchmark, this is the result of my running:
```
$ go test -bench=BenchmarkIndexValueEntry_Add
goos: windows
goarch: amd64
pkg: github.com/cortexproject/cortex/pkg/ingester/index
BenchmarkIndexValueEntry_Add/shard-12            2527686              3076 ns/op
BenchmarkIndexValueEntry_Add/plain-12            1000000            186148 ns/op
PASS
ok      github.com/cortexproject/cortex/pkg/ingester/index      194.862s
```

If my benchmark is written without problems, after the shard, the insertion speed is increased by 80 times

**Comment by bboreham** [comment]
> Each series contains only one sample and about 10 labels.

Cortex (and Prometheus) is designed to handle timeseries that have many values over hours; if you have only one sample in a series then metadata (labels, etc) dominates storage and processing time. 

Don't you find that the IO costs of putting this on disk are enormous?  I suspect an entirely different storage scheme would be needed to make this style of data manageable.

**Comment by storyicon** [comment]
If I configured `scrape_interval` as 60s in prometheus, which means that only one sample is generated every 60 seconds for these series. If prometheus requests cortex every 60 seconds (for remote write), this is what we see:
```
Each series contains only one sample and about 10 labels
```
This is common in my environment.

**Comment by bboreham** [comment]
Based on further discussion on Slack, I think you are focused on the time Cortex sees a lot of series for the first time, e.g. after an ingester is restarted, or if a large (tens of millions of series) Prometheus starts sending for the first time.

This is a valid concern, though perhaps one that has slipped past other large users.  Possibly other users are scaled more horizontally - the copying effect you describe will increase exponentially as numbers of series in one shard grows.

I think the same effect should be observable in the ingester hand-over process, where millions of series will be created in a single transaction.

**Comment by storyicon** [comment]
@bboreham  Sorry, I was busy with some other things a few days ago, and today I see your reply here.

In order to make the comparison clearer, I re-tested under the same conditions:

1. Before fixing
image

    It can be seen that a large number of prometheus latency lasted for about 2h, this problem is fatal enough.

2. After fixing
image

    The accumulation time of fingerprints only consumes about 10 minutes, and the short latency is also acceptable.

The test data used above comes from our production environment (not any special stress test data), which only accounts for about 20% of the total. (This is also the data source I have been using before)

When most fingerprints already exist in memory, the speed of both has been greatly improved, but there is no doubt that the revised version will always be a little faster. If a new batch of data is accessed at this time, the fixed version will be able to digest them quickly.
But anyway, there is a two-hour accumulation before fixing, which is basically unavailable.
I admit that through scale-out can help reduce this "start-up" time, but scale-out is not free for cortex, for example: _In order to reduce data skew, we need to turn on ShardByAllLabels, which will cause Querier to query all Ingester nodes when querying_.
The ability to expand horizontally does not conflict with improving the performance of every single Ingester. This fix may only show its advantages when the amount of data is large enough, but I think with the development of microservices, the amount of data in each company is expanding rapidly, a large amount of prometheus latency may cause users to worry about cortex performance.

**Pull request closed without merging**